PeriodogramAnalysis/test_power_density_statistics.py

```diff
@@ -0,0 +1,237 @@
+from typing import Generator, Any
+
+import numpy as np
+import pytest
+from collections import defaultdict
+from miv.core.datatype import Signal
+from power_density_statistics import (
+    SpectrumAnalysisBase,
+    SpectrumAnalysisPeriodogram,
+    SpectrumAnalysisWelch,
+)
+
+
+def mock_signal_generator() -> Generator:
+    timestamps = np.linspace(0, 10, 1000, endpoint=False)
+    signal = (
+        np.sin(2 * np.pi * 5 * timestamps)
+        + np.sin(2 * np.pi * 10 * timestamps)
+        + np.sin(2 * np.pi * 15 * timestamps)
+        + np.sin(2 * np.pi * 20 * timestamps)
+    )
+
+    data1 = np.array([signal, signal]).T
+    signal1 = Signal(data=data1, timestamps=timestamps, rate=100)
+    yield signal1
+
+    data2 = np.array([signal, signal]).T
+    signal2 = Signal(data=data2, timestamps=timestamps, rate=100)
+    yield signal2
+
+
+def test_SpectrumAnalysisBase_compute_psd_not_implemented() -> None:
+    """
+    This test is used to test if no subclass is used, will it pop up proper error.
+    """
+    analyzer = SpectrumAnalysisBase()
+    signal = next(mock_signal_generator())
+    psd_dict: dict[int, dict[str, Any]] = {}
+    with pytest.raises(NotImplementedError) as exc_info:
+        analyzer.compute_psd(signal, psd_dict)
+    assert str(exc_info.value) == (
+        "The compute_psd method is not implemented in the base class. "
+        "This base class is not intended for standalone use. Please use a subclass "
+        "such as SpectrumAnalysisWelch, SpectrumAnalysisPeriodogram, or SpectrumAnalysisMultitaper."
+    )
+
+
+def test_SpectrumAnalysisBase_call(mocker) -> None:
```

Comment:
Separate out the test.

Test to check parameters: For this, you need to test negatives: integer window length, non-negative band, ordered band (band[0] < band[1]) etc. If your function runs with wrong parameters, your code is not safe. Here, you are just testing if the default value is the same, which doesn't give any useful test for two reasons: your test depends on your implementation which means your test needs to be adapted to your implementation. It is not how the test should be written. Secondly, if this test fails, it just means your test is wrong, not your implementation. Think about the purpose of writing unittest.

Reply:
I have added more negative tests for parameters: https://github.com/Qixian-Wang/NEF/blob/main/PeriodogramAnalysis/spectrogram_analysis.py#L38-L55.     I am not sure if it is too much though.

Reply:
Revision to this test_file is in this commit: [8dcc53a]
Revisions to the rest is here: [1a0b4a9]